testing/integration/agent_long_test.go

```diff
@@ -0,0 +1,282 @@
+	apachePackage.PolicyID = policyResp.ID
+	apachePackage.Namespace = "default"
+	apachePackage.Name = fmt.Sprintf("system-long-test-%s", policyUUIDApache)
+	apachePackage.Vars = map[string]interface{}{}
+
+	runner.T().Logf("Installing fleet package....")
+	_, err = runner.info.KibanaClient.InstallFleetPackage(ctx, apachePackage)
+	require.NoError(runner.T(), err, "error creating fleet package")
+
+}
+
+func (runner *ExtendedRunner) TestHandleLeak() {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Hour)
+	defer cancel()
+
+	testRuntime := os.Getenv("LONG_TEST_RUNTIME")
+	if testRuntime == "" {
+		testRuntime = "5m"
+	}
+
+	// because we need to separately fetch the PIDs, wait until everything is healthy before we look for running beats
+	require.Eventually(runner.T(), func() bool {
+		allHealthy := true
+		status, err := runner.agentFixture.ExecStatus(context.Background())
+		require.NoError(runner.T(), err)
+		for _, comp := range status.Components {
+			if comp.State != int(cproto.State_HEALTHY) {
+				allHealthy = false
+			}
+		}
+		return allHealthy
+	}, time.Minute*3, time.Second*20)
+
+	procs, err := sysinfo.Processes()
+	require.NoError(runner.T(), err)
+
+	handles := []processWatcher{}
+
+	// track running beats
+	// the `last 30s` metrics tend to report gauges, which we can't use for calculating a derivative.
+	// so separately fetch the PIDs
+	for _, proc := range procs {
+		info, err := proc.Info()
+		require.NoError(runner.T(), err)
+		if strings.Contains(info.Name, "beat") || strings.Contains(info.Name, "elastic-agent") {
+			handle, err := sysinfo.Process(proc.PID())
+			require.NoError(runner.T(), err)
+			handlesReg := new(regression.Regression)
+			handlesReg.SetObserved(fmt.Sprintf("%s handle usage", info.Name))
+			handlesReg.SetVar(0, "time")
+
+			runner.T().Logf("created handle watcher for %s (%d)", info.Name, proc.PID())
+			handles = append(handles, processWatcher{handle: handle, pid: proc.PID(), name: info.Name, regHandles: handlesReg})
+		}
+	}
+
+	testDuration, err := time.ParseDuration(testRuntime)
+	require.NoError(runner.T(), err)
+
+	timer := time.NewTimer(testDuration)
+
+	// time to perform a health check
+	ticker := time.Tick(time.Second * 10)
+
+	done := false
+	start := time.Now()
+	for !done {
+		select {
+		case <-timer.C:
+			done = true
+		case <-ticker:
+			err := runner.agentFixture.IsHealthy(ctx)
+			require.NoError(runner.T(), err)
+			// for each running process, collect memory and handles
+			for _, handle := range handles {
+
+				ohc, ok := handle.handle.(types.OpenHandleCounter)
+				if ok {
+					handleCount, err := ohc.OpenHandleCount()
+					require.NoError(runner.T(), err)
+					handle.regHandles.Train(regression.DataPoint(float64(handleCount), []float64{time.Since(start).Seconds()}))
+				}
+
+			}
+		}
+	}
+
+	// we're measuring the handle usage as y=mx+b
+	// if the slope is increasing above a certain rate, fail the test
+	handleSlopeFailure := float64(1)
```

Comment:
What is the slope you measure when the 8.11 leak is happening?

What is the lowest slope that isn't flaky? Should we experiment to find the slowest handle leak we can actively detect?

You probably need to leak handles on purpose in a controlled way to test this.

I don't want us to merge this and think "we will never have a handle leak again", and then end up releasing a leak that is just slightly slower than this threshold.

Reply:
Yah, there's room to tweak that number. The lowest number I've seen 8.11 return for the slope is around 500, but I pulled that number based on the fact that all of the functioning releases are less than 1. It generally ranges around `0.001` to `1.0e-6` depending on how long the test is run for. We can start with a lower number and see how flaky that is.

Reply:
Yeah what worries me is setting this slightly too high and failing to detect a leak.

It sounds like we could go down to the 0.25-0.5 range.

Reply:
Is this value considering the entire runtime? As in is it telling us that over the entire 20m runtime of the test the slope was 0.5? As in telling us the number was trending up the entire time? In that case any number >0 indicates a leak and that is why on average it is essentially zero.

Possibly we could get away with 0.1 but we'd need to test to be sure.

Reply:
So, we're just asking the OS for current open handles, which means that if a handle was opened right at the beginning and never closed, we'll see it one minute later, and so on.

So, a lot of the results are basically rounding errors, `1.0e-6`, and so on. I think a lot of the "higher" numbers (like `0.003`) come from filebeat, and I assume are related to it chugging through log files while the numbers are being collected, resulting in a slight positive curve?